### PR TITLE
feat(engine): [YW-140] generic declarative RestConnector (HTTP/JSON any-source)

### PR DESCRIPTION
**Pull request by youhaowei**
## Summary

- Adds `@dashframe/connector-rest` — a new package implementing `RestConnector`, a generic declarative HTTP/JSON connector that interprets `DataSource.config` (endpoint, method, authRef, pagination, rowPath, fieldMap) to pull from any HTTP/JSON source. Config IS the mod — no executable code in config.
- Implements all four pagination strategies: **offset**, **cursor**, **page-number**, and **link-header** (RFC 5988). Unsupported strategy strings are logged as declarative-ceiling gaps for the v0.4 code-plugin tier.
- `authRef` resolves server-side through the `SecretResolver` scoped lease (vault's `withSecret`). Vault-absent + authRef-present → throw before any fetch (fail-closed). Plaintext never in config, formData, or any public interface.
- `connect()` performs a live single-page test fetch (real-world side-effect — cannot ride a preview transaction).
- `query()` returns `{ arrowBuffer (base64), fieldIds, fields }` — fully serializable over IPC. Renderer materializes the DataFrame from arrowBuffer.
- rowPath dot-path extraction and fieldMap key renaming, with Arrow IPC output via `apache-arrow` `tableFromArrays`/`tableToIPC`.

Tracked internally as YW-140.

## Test plan

- [x] 24/24 tests pass (`bunx vitest run` in `packages/connector-rest`)
- [x] Full turbo typecheck clean (46/46 tasks pass)
- [x] Offset pagination: two-page fetch, stops on empty page
- [x] Cursor pagination: follows `next_cursor` until absent
- [x] Page-number pagination: increments page, stops on partial page
- [x] Link-header pagination: follows `Link: rel="next"` header, stops when absent
- [x] rowPath + fieldMap: nested extraction + key rename confirmed
- [x] authRef vault resolution: Bearer token injected into fetch headers via vault's withSecret; plaintext never in result or formData
- [x] Fail-closed: SecretResolver rejection → throw, fetch never called
- [x] Public endpoints (noopResolver): no Authorization header sent

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR introduces `@dashframe/connector-rest`, a new declarative HTTP/JSON connector package that drives pagination, auth, row extraction, and Arrow serialization entirely from a `RestConnectorConfig` object — no executable code in config.

- Implements four pagination strategies (offset, cursor, page-number, link-header/RFC 5988) with budget-aware early-stop, a same-origin SSRF guard on Link-header redirects, and fail-closed auth enforcement when `authRef` is configured.
- Serializes results as Apache Arrow IPC (base64) with per-cell type coercion aligned to `inferStringColumnType`, null-prototype row objects to block prototype pollution, and fieldMap collision detection; 24 tests cover all strategies, security edges, and coercion edge-cases.
</details>

<details><summary><h3>Confidence Score: 3/5</h3></summary>

Safe to merge after addressing the cursor `"next"` probe bug; the rest of the connector is well-guarded and thoroughly tested.

The cursor pagination auto-detection probes `"next"` as a fallback field name. On APIs that place the next-page URL in a response body `next` field (a very common pattern), the connector silently duplicates page-1 rows until the row budget is exhausted — no error is raised, so callers receive wrong data without knowing it.

packages/connector-rest/src/connector.ts — specifically the `resolveNextCursor` function and the `fetchPage` body-less implementation for non-GET methods.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/connector-rest/src/connector.ts | Core connector implementation; contains the resolveNextCursor P1 bug (probing "next" as cursor name causes silent row duplication on DRF-style APIs) and a P2 gap where POST/PUT/PATCH methods have no body support. |
| packages/connector-rest/src/connector.test.ts | Comprehensive 857-line test suite covering all four pagination strategies, auth/vault resolution, SSRF guard, prototype-pollution defence, limit pushdown, type coercion, and fieldMap collision detection. |
| packages/connector-rest/src/types.ts | Declarative config types for RestConnectorConfig and PaginationStrategy; well-documented with no issues. |
| packages/connector-rest/src/index.ts | Re-exports public API surface from connector and types; no issues. |
| packages/connector-rest/package.json | New package manifest for @dashframe/connector-rest; dependencies and scripts look correct. |
| packages/connector-rest/vitest.config.ts | Vitest config with workspace aliases for engine and secret-vault; coverage thresholds set at 80%; no issues. |

</details>

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Client
    participant RestConnector
    participant SecretVault as SecretResolver / Vault
    participant RemoteAPI

    Client->>RestConnector: query(databaseId, tableId, options?)
    RestConnector->>RestConnector: assertFieldMapNoCollision(fieldMap)
    RestConnector->>SecretVault: auth(callback)
    SecretVault-->>RestConnector: token (throws if authRef set and token empty)
    RestConnector->>RestConnector: authHeaders(token) → headers map

    loop fetchAllPages (offset / page-number / cursor / link-header)
        RestConnector->>RemoteAPI: "fetch(url, { method, headers })"
        RemoteAPI-->>RestConnector: JSON response
        RestConnector->>RestConnector: extractRows(data, rowPath)
        RestConnector->>RestConnector: check budget (sliceOffset + limit)
    end

    RestConnector->>RestConnector: filter and applyFieldMap(rows)
    RestConnector->>RestConnector: slice(sliceOffset, sliceOffset+limit)
    RestConnector->>RestConnector: inferFieldsFromRows(limitedRows)
    RestConnector->>RestConnector: coerceValueToType() per cell
    RestConnector->>RestConnector: tableFromArrays() → tableToIPC() → base64
    RestConnector-->>Client: "{ arrowBuffer, fieldIds, fields }"
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Client
    participant RestConnector
    participant SecretVault as SecretResolver / Vault
    participant RemoteAPI

    Client->>RestConnector: query(databaseId, tableId, options?)
    RestConnector->>RestConnector: assertFieldMapNoCollision(fieldMap)
    RestConnector->>SecretVault: auth(callback)
    SecretVault-->>RestConnector: token (throws if authRef set and token empty)
    RestConnector->>RestConnector: authHeaders(token) → headers map

    loop fetchAllPages (offset / page-number / cursor / link-header)
        RestConnector->>RemoteAPI: "fetch(url, { method, headers })"
        RemoteAPI-->>RestConnector: JSON response
        RestConnector->>RestConnector: extractRows(data, rowPath)
        RestConnector->>RestConnector: check budget (sliceOffset + limit)
    end

    RestConnector->>RestConnector: filter and applyFieldMap(rows)
    RestConnector->>RestConnector: slice(sliceOffset, sliceOffset+limit)
    RestConnector->>RestConnector: inferFieldsFromRows(limitedRows)
    RestConnector->>RestConnector: coerceValueToType() per cell
    RestConnector->>RestConnector: tableFromArrays() → tableToIPC() → base64
    RestConnector-->>Client: "{ arrowBuffer, fieldIds, fields }"
```

</a>
</details>

<details><summary><h3>Comments Outside Diff (2)</h3></summary>

1. `packages/connector-rest/src/connector.ts`, line 751-806 ([link](https://github.com/youhaowei/dashframe/blob/5b85058ba81505870c01e8e0ee6a8eb1f237f500/packages/connector-rest/src/connector.ts#L751-L806)) 

   <a href="#">P1</a> **Unbounded pagination loops — no max-pages or max-rows guard**

   The `cursor` and `link-header` cases have no upper bound on iterations. If an API always returns a non-empty cursor or a `Link: rel="next"` header (e.g., a bug on the remote end, a circular ref, or a misconfigured `cursorPath`), `fetchAllPages` will loop indefinitely and the `query()` call will never return. The `offset` and `page-number` cases are partially protected by the `rows.length === 0` check but are similarly unbounded for row accumulation.

   A simple guard — e.g., `const MAX_PAGES = 1_000; let pages = 0;` at the top of `fetchAllPages` with `if (++pages > MAX_PAGES) { console.warn(...); break; }` at the top of each loop — would prevent this without changing normal behaviour.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: packages/connector-rest/src/connector.ts
   Line: 751-806

   Comment:
   **Unbounded pagination loops — no max-pages or max-rows guard**

   The `cursor` and `link-header` cases have no upper bound on iterations. If an API always returns a non-empty cursor or a `Link: rel="next"` header (e.g., a bug on the remote end, a circular ref, or a misconfigured `cursorPath`), `fetchAllPages` will loop indefinitely and the `query()` call will never return. The `offset` and `page-number` cases are partially protected by the `rows.length === 0` check but are similarly unbounded for row accumulation.

   A simple guard — e.g., `const MAX_PAGES = 1_000; let pages = 0;` at the top of `fetchAllPages` with `if (++pages > MAX_PAGES) { console.warn(...); break; }` at the top of each loop — would prevent this without changing normal behaviour.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fconnector-rest%2Fsrc%2Fconnector.ts%0ALine%3A%20751-806%0A%0AComment%3A%0A**Unbounded%20pagination%20loops%20%E2%80%94%20no%20max-pages%20or%20max-rows%20guard**%0A%0AThe%20%60cursor%60%20and%20%60link-header%60%20cases%20have%20no%20upper%20bound%20on%20iterations.%20If%20an%20API%20always%20returns%20a%20non-empty%20cursor%20or%20a%20%60Link%3A%20rel%3D%22next%22%60%20header%20%28e.g.%2C%20a%20bug%20on%20the%20remote%20end%2C%20a%20circular%20ref%2C%20or%20a%20misconfigured%20%60cursorPath%60%29%2C%20%60fetchAllPages%60%20will%20loop%20indefinitely%20and%20the%20%60query%28%29%60%20call%20will%20never%20return.%20The%20%60offset%60%20and%20%60page-number%60%20cases%20are%20partially%20protected%20by%20the%20%60rows.length%20%3D%3D%3D%200%60%20check%20but%20are%20similarly%20unbounded%20for%20row%20accumulation.%0A%0AA%20simple%20guard%20%E2%80%94%20e.g.%2C%20%60const%20MAX_PAGES%20%3D%201_000%3B%20let%20pages%20%3D%200%3B%60%20at%20the%20top%20of%20%60fetchAllPages%60%20with%20%60if%20%28%2B%2Bpages%20%3E%20MAX_PAGES%29%20%7B%20console.warn%28...%29%3B%20break%3B%20%7D%60%20at%20the%20top%20of%20each%20loop%20%E2%80%94%20would%20prevent%20this%20without%20changing%20normal%20behaviour.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=149&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22youhaowei%2Fdashframe%22%20on%20the%20existing%20branch%20%22charixandra%2Fyw-140-generic-rest-connector%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22charixandra%2Fyw-140-generic-rest-connector%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fconnector-rest%2Fsrc%2Fconnector.ts%0ALine%3A%20751-806%0A%0AComment%3A%0A**Unbounded%20pagination%20loops%20%E2%80%94%20no%20max-pages%20or%20max-rows%20guard**%0A%0AThe%20%60cursor%60%20and%20%60link-header%60%20cases%20have%20no%20upper%20bound%20on%20iterations.%20If%20an%20API%20always%20returns%20a%20non-empty%20cursor%20or%20a%20%60Link%3A%20rel%3D%22next%22%60%20header%20%28e.g.%2C%20a%20bug%20on%20the%20remote%20end%2C%20a%20circular%20ref%2C%20or%20a%20misconfigured%20%60cursorPath%60%29%2C%20%60fetchAllPages%60%20will%20loop%20indefinitely%20and%20the%20%60query%28%29%60%20call%20will%20never%20return.%20The%20%60offset%60%20and%20%60page-number%60%20cases%20are%20partially%20protected%20by%20the%20%60rows.length%20%3D%3D%3D%200%60%20check%20but%20are%20similarly%20unbounded%20for%20row%20accumulation.%0A%0AA%20simple%20guard%20%E2%80%94%20e.g.%2C%20%60const%20MAX_PAGES%20%3D%201_000%3B%20let%20pages%20%3D%200%3B%60%20at%20the%20top%20of%20%60fetchAllPages%60%20with%20%60if%20%28%2B%2Bpages%20%3E%20MAX_PAGES%29%20%7B%20console.warn%28...%29%3B%20break%3B%20%7D%60%20at%20the%20top%20of%20each%20loop%20%E2%80%94%20would%20prevent%20this%20without%20changing%20normal%20behaviour.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=149&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fconnector-rest%2Fsrc%2Fconnector.ts%0ALine%3A%20751-806%0A%0AComment%3A%0A**Unbounded%20pagination%20loops%20%E2%80%94%20no%20max-pages%20or%20max-rows%20guard**%0A%0AThe%20%60cursor%60%20and%20%60link-header%60%20cases%20have%20no%20upper%20bound%20on%20iterations.%20If%20an%20API%20always%20returns%20a%20non-empty%20cursor%20or%20a%20%60Link%3A%20rel%3D%22next%22%60%20header%20%28e.g.%2C%20a%20bug%20on%20the%20remote%20end%2C%20a%20circular%20ref%2C%20or%20a%20misconfigured%20%60cursorPath%60%29%2C%20%60fetchAllPages%60%20will%20loop%20indefinitely%20and%20the%20%60query%28%29%60%20call%20will%20never%20return.%20The%20%60offset%60%20and%20%60page-number%60%20cases%20are%20partially%20protected%20by%20the%20%60rows.length%20%3D%3D%3D%200%60%20check%20but%20are%20similarly%20unbounded%20for%20row%20accumulation.%0A%0AA%20simple%20guard%20%E2%80%94%20e.g.%2C%20%60const%20MAX_PAGES%20%3D%201_000%3B%20let%20pages%20%3D%200%3B%60%20at%20the%20top%20of%20%60fetchAllPages%60%20with%20%60if%20%28%2B%2Bpages%20%3E%20MAX_PAGES%29%20%7B%20console.warn%28...%29%3B%20break%3B%20%7D%60%20at%20the%20top%20of%20each%20loop%20%E2%80%94%20would%20prevent%20this%20without%20changing%20normal%20behaviour.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=149&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a>

2. `packages/connector-rest/src/connector.ts`, line 966-983 ([link](https://github.com/youhaowei/dashframe/blob/5b85058ba81505870c01e8e0ee6a8eb1f237f500/packages/connector-rest/src/connector.ts#L966-L983)) 

   <a href="#">P2</a> **Full remote dataset fetched before `QueryOptions` row limit is applied**

   `fetchAllPages` downloads every page into `rawRows` before `options?.pagination?.limit` slices the result. If the remote API serves thousands of pages and the UI requests only 50 rows, the connector still fetches and materializes the full dataset in memory. For the v0.1 scope this may be acceptable, but the current `options.pagination` slice gives callers no way to avoid the full download — they cannot tell from the interface that specifying a small `limit` won't reduce network traffic.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: packages/connector-rest/src/connector.ts
   Line: 966-983

   Comment:
   **Full remote dataset fetched before `QueryOptions` row limit is applied**

   `fetchAllPages` downloads every page into `rawRows` before `options?.pagination?.limit` slices the result. If the remote API serves thousands of pages and the UI requests only 50 rows, the connector still fetches and materializes the full dataset in memory. For the v0.1 scope this may be acceptable, but the current `options.pagination` slice gives callers no way to avoid the full download — they cannot tell from the interface that specifying a small `limit` won't reduce network traffic.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fconnector-rest%2Fsrc%2Fconnector.ts%0ALine%3A%20966-983%0A%0AComment%3A%0A**Full%20remote%20dataset%20fetched%20before%20%60QueryOptions%60%20row%20limit%20is%20applied**%0A%0A%60fetchAllPages%60%20downloads%20every%20page%20into%20%60rawRows%60%20before%20%60options%3F.pagination%3F.limit%60%20slices%20the%20result.%20If%20the%20remote%20API%20serves%20thousands%20of%20pages%20and%20the%20UI%20requests%20only%2050%20rows%2C%20the%20connector%20still%20fetches%20and%20materializes%20the%20full%20dataset%20in%20memory.%20For%20the%20v0.1%20scope%20this%20may%20be%20acceptable%2C%20but%20the%20current%20%60options.pagination%60%20slice%20gives%20callers%20no%20way%20to%20avoid%20the%20full%20download%20%E2%80%94%20they%20cannot%20tell%20from%20the%20interface%20that%20specifying%20a%20small%20%60limit%60%20won't%20reduce%20network%20traffic.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=149&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22youhaowei%2Fdashframe%22%20on%20the%20existing%20branch%20%22charixandra%2Fyw-140-generic-rest-connector%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22charixandra%2Fyw-140-generic-rest-connector%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fconnector-rest%2Fsrc%2Fconnector.ts%0ALine%3A%20966-983%0A%0AComment%3A%0A**Full%20remote%20dataset%20fetched%20before%20%60QueryOptions%60%20row%20limit%20is%20applied**%0A%0A%60fetchAllPages%60%20downloads%20every%20page%20into%20%60rawRows%60%20before%20%60options%3F.pagination%3F.limit%60%20slices%20the%20result.%20If%20the%20remote%20API%20serves%20thousands%20of%20pages%20and%20the%20UI%20requests%20only%2050%20rows%2C%20the%20connector%20still%20fetches%20and%20materializes%20the%20full%20dataset%20in%20memory.%20For%20the%20v0.1%20scope%20this%20may%20be%20acceptable%2C%20but%20the%20current%20%60options.pagination%60%20slice%20gives%20callers%20no%20way%20to%20avoid%20the%20full%20download%20%E2%80%94%20they%20cannot%20tell%20from%20the%20interface%20that%20specifying%20a%20small%20%60limit%60%20won't%20reduce%20network%20traffic.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=149&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fconnector-rest%2Fsrc%2Fconnector.ts%0ALine%3A%20966-983%0A%0AComment%3A%0A**Full%20remote%20dataset%20fetched%20before%20%60QueryOptions%60%20row%20limit%20is%20applied**%0A%0A%60fetchAllPages%60%20downloads%20every%20page%20into%20%60rawRows%60%20before%20%60options%3F.pagination%3F.limit%60%20slices%20the%20result.%20If%20the%20remote%20API%20serves%20thousands%20of%20pages%20and%20the%20UI%20requests%20only%2050%20rows%2C%20the%20connector%20still%20fetches%20and%20materializes%20the%20full%20dataset%20in%20memory.%20For%20the%20v0.1%20scope%20this%20may%20be%20acceptable%2C%20but%20the%20current%20%60options.pagination%60%20slice%20gives%20callers%20no%20way%20to%20avoid%20the%20full%20download%20%E2%80%94%20they%20cannot%20tell%20from%20the%20interface%20that%20specifying%20a%20small%20%60limit%60%20won't%20reduce%20network%20traffic.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=149&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a>
</details>

<sub>Reviews (3): Last reviewed commit: ["fix(connector-rest): address security + ..."](https://github.com/youhaowei/dashframe/commit/416a9ed7045865a7ffbaf8ab67d1861d46ee3245) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38149077)</sub>

### CODE REVIEW

File: packages/connector-rest/src/connector.ts [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+
+  while (nextUrl) {
+    const { data, response } = await fetchPage(nextUrl, method, headers);
+    allRows.push(...extractRows(data, config.rowPath));
+    const linkHeader = response.headers.get("Link");
+    nextUrl = linkHeader ? parseLinkNext(linkHeader) : undefined;
+  }
+
+  return allRows;
+}
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-rest/src/connector.ts:504]: Fixed in 416a9ed7. `fetchLinkHeaderPages` now anchors a same-origin guard to the immutable config endpoint origin: the vault-resolved Bearer token is forwarded ONLY to a same-origin `rel="next"` URL. A cross-origin (or unresolvable) next-link stops pagination with a warning — the credential is never sent off-origin. The origin is computed once from `config.endpoint` and never walked forward by the link chain. Test added: cross-origin next-link receives no request and no token.

File: packages/connector-rest/src/connector.ts [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+   * Credentials are resolved via `this.auth` — no credential argument.
+   */
+  async query(
+    _databaseId: string,
+    tableId: UUID,
+    options?: QueryOptions,
+  ): Promise<ConnectorQueryResult> {
+    return this.auth(async (token) => {
+      const headers: Record<string, string> = {};
+      if (token) headers["Authorization"] = `Bearer ${token}`;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/connector-rest/src/connector.ts]: Fixed in 416a9ed7. New `#authHeaders` throws BEFORE any fetch when `config.authRef` is set but the resolved token is empty — never issues an unauthenticated request as if it were fine. Both `connect()` and `query()` route through it. Tests added: authRef-set-but-empty-token throws before fetch in both paths (fetch never called).
